File: services/user-service/src/modules/users/events/kafka-user-events.publisher.ts
import { Injectable, Logger, OnModuleDestroy, OnModuleInit } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Kafka, Producer } from 'kafkajs';
import { UserEventsPublisher, UserRegisteredEventPayload } from './user-events.publisher';

@Injectable()
export class KafkaUserEventsPublisher implements UserEventsPublisher, OnModuleInit, OnModuleDestroy {
  private readonly logger = new Logger(KafkaUserEventsPublisher.name);
  private readonly producer: Producer | null;
  private readonly topic: string;

  constructor(private readonly configService: ConfigService) {
    const brokers = this.configService
      .get<string>('KAFKA_BROKERS', '')
      .split(',')
      .map((value) => value.trim())
      .filter(Boolean);

    this.topic = this.configService.get<string>('KAFKA_USER_TOPIC', 'user.registered');
    const isEnabled = this.configService.get<boolean>('KAFKA_ENABLED', false) && brokers.length > 0;

    if (!isEnabled) {
      this.producer = null;
      return;
    }

    const kafka = new Kafka({
      clientId: this.configService.get<string>('KAFKA_CLIENT_ID', 'user-service'),
      brokers
    });
    this.producer = kafka.producer();
  }

  async onModuleInit(): Promise<void> {
    if (!this.producer) {
      return;
    }
    await this.producer.connect();
  }

  async onModuleDestroy(): Promise<void> {
    if (!this.producer) {
      return;
    }
    await this.producer.disconnect();
  }

  async publishUserRegistered(event: UserRegisteredEventPayload): Promise<void> {
    if (!this.producer) {
      this.logger.debug('Kafka is disabled, skip publishing user.registered event');
      return;
    }

    await this.producer.send({
      topic: this.topic,
      messages: [
        {
          key: event.userId,
          value: JSON.stringify(event)
        }
      ]
    });
  }
}

File: services/user-service/internal/events/kafka_user_events_publisher.go
package events

import (
	"context"
	"encoding/json"
	"time"

	"user-service-go/internal/config"

	"github.com/segmentio/kafka-go"
	"go.uber.org/zap"
)

type KafkaUserEventsPublisher struct {
	logger  *zap.Logger
	topic   string
	enabled bool
	writer  *kafka.Writer
}

func NewKafkaUserEventsPublisher(cfg config.Config, logger *zap.Logger) UserEventsPublisher {
	enabled := cfg.KafkaEnabled && len(cfg.KafkaBrokers) > 0
	if !enabled {
		logger.Debug("kafka disabled, user.registered publishing skipped")
		return &NoopUserEventsPublisher{}
	}

	writer := &kafka.Writer{
		Addr:         kafka.TCP(cfg.KafkaBrokers...),
		Topic:        cfg.KafkaUserTopic,
		RequiredAcks: kafka.RequireOne,
		Balancer:     &kafka.LeastBytes{},
		BatchTimeout: 10 * time.Millisecond,
		Async:        false,
		Transport: &kafka.Transport{
			ClientID: cfg.KafkaClientID,
		},
	}

	return &KafkaUserEventsPublisher{
		logger:  logger,
		topic:   cfg.KafkaUserTopic,
		enabled: true,
		writer:  writer,
	}
}

func (k *KafkaUserEventsPublisher) PublishUserRegistered(ctx context.Context, event UserRegisteredEventPayload) error {
	if !k.enabled || k.writer == nil {
		return nil
	}

	payload, err := json.Marshal(event)
	if err != nil {
		return err
	}

	return k.writer.WriteMessages(ctx, kafka.Message{
		Key:   []byte(event.UserID),
		Value: payload,
	})
}

func (k *KafkaUserEventsPublisher) Close(ctx context.Context) error {
	if !k.enabled || k.writer == nil {
		return nil
	}
	return k.writer.Close()
}
